File: bin/dww/bindings/niche/darkwow.lua
-- DarkWow Wallet — LuaJIT FFI binding
--   local darkwow = require("darkwow")
--   local w = darkwow.open("keys.toml", "wallet-1", "testnet")
--   print("balance=" .. w:balance() .. " caps=" .. w:capCount())
--   w:close()

local ffi = require("ffi")

ffi.cdef([[
    typedef struct WalletHandle WalletHandle;
    const char* dwow_wallet_version(void);
    WalletHandle* dwow_wallet_open(const char* keys_path, const char* section, const char* network);
    void dwow_wallet_free(WalletHandle* handle);
    int dwow_wallet_scan_block_json(WalletHandle* handle, const char* block_json);
    int dwow_wallet_cap_count(WalletHandle* handle);
    uint64_t dwow_wallet_balance(WalletHandle* handle);
    uint64_t dwow_wallet_chain_height(WalletHandle* handle);
]])

local lib = ffi.load(os.getenv("DARKWOW_LIB") or "libdwow_wallet")

local DarkWow = {}
DarkWow.__index = DarkWow

function DarkWow.open(keysPath, section, network)
    network = network or "testnet"
    local handle = lib.dwow_wallet_open(keysPath, section, network)
    if handle == nil then error("dwow_wallet_open failed") end
    return setmetatable({ handle = handle }, DarkWow)
end

function DarkWow:close()
    if self.handle then lib.dwow_wallet_free(self.handle); self.handle = nil end
end

function DarkWow:scanBlock(json)
    return lib.dwow_wallet_scan_block_json(self.handle, json)
end

function DarkWow:capCount() return lib.dwow_wallet_cap_count(self.handle) end
function DarkWow:balance() return tonumber(lib.dwow_wallet_balance(self.handle)) end
function DarkWow:chainHeight() return tonumber(lib.dwow_wallet_chain_height(self.handle)) end

DarkWow.version = function() return ffi.string(lib.dwow_wallet_version()) end

return DarkWow

File: bin/dww/bindings/niche/darkwow.ml
(* DarkWow Wallet — OCaml ctypes binding
   (* Compile: ocamlfind opt -package ctypes.foreign -linkpkg darkwow.ml *)
   open DarkWow
   let () =
     let w = open_wallet "keys.toml" "wallet-1" "testnet" in
     let n = scan_block w block_json in
     Printf.printf "scanned %d outputs, balance=%Ld\n" n (balance w);
     close_wallet w
*)

open Ctypes
open Foreign

type wallet_handle = unit ptr
let wallet_handle : wallet_handle typ = ptr void

let version_fn = foreign "dwow_wallet_version" (void @-> returning string)

let open_fn = foreign "dwow_wallet_open"
    (string @-> string @-> string @-> returning wallet_handle)

let free_fn = foreign "dwow_wallet_free" (wallet_handle @-> returning void)

let scan_fn = foreign "dwow_wallet_scan_block_json"
    (wallet_handle @-> string @-> returning int)

let cap_count_fn = foreign "dwow_wallet_cap_count"
    (wallet_handle @-> returning int)

let balance_fn = foreign "dwow_wallet_balance"
    (wallet_handle @-> returning int64_t)

let chain_height_fn = foreign "dwow_wallet_chain_height"
    (wallet_handle @-> returning int64_t)

let open_wallet kp sec net = open_fn kp sec net
let close_wallet = free_fn
let scan_block h json = scan_fn h json
let cap_count h = cap_count_fn h
let balance h = Int64.to_int (balance_fn h)
let chain_height h = Int64.to_int (chain_height_fn h)
let version () = version_fn ()
